lib/rmg/TreasurePlacer.cpp

```diff
@@ -0,0 +1,795 @@
+{
+	int mapMonsterStrength = map.getMapGenOptions().getMonsterStrength();
+	int monsterStrength = zone.zoneMonsterStrength + mapMonsterStrength - 1; //array index from 0 to 4
+	
+	static int minGuardedValues[] = { 6500, 4167, 3000, 1833, 1333 };
+	minGuardedValue = minGuardedValues[monsterStrength];
+	
+	auto valueComparator = [](const CTreasureInfo & lhs, const CTreasureInfo & rhs) -> bool
+	{
+		return lhs.max > rhs.max;
+	};
+	
+	auto restoreZoneLimits = [](const std::vector<ObjectInfo*> & treasurePile)
+	{
+		for(auto * oi : treasurePile)
+		{
+			oi->maxPerZone++;
+		}
+	};
+	
+	//place biggest treasures first at large distance, place smaller ones inbetween
+	auto treasureInfo = zone.getTreasureInfo();
+	boost::sort(treasureInfo, valueComparator);
+	
+	//sort treasures by ascending value so we can stop checking treasures with too high value
+	boost::sort(possibleObjects, [](const ObjectInfo& oi1, const ObjectInfo& oi2) -> bool
+	{
+		return oi1.value < oi2.value;
+	});
+	
+	int totalDensity = 0;
+	for (auto t : treasureInfo)
+	{
+		//discard objects with too high value to be ever placed
+		vstd::erase_if(possibleObjects, [t](const ObjectInfo& oi) -> bool
+		{
+			return oi.value > t.max;
+		});
+		
+		totalDensity += t.density;
+		
+		//treasure density is inversely proportional to zone size but must be scaled back to map size
+		//also, normalize it to zone count - higher count means relatively smaller zones
+		
+		//this is squared distance for optimization purposes
+		const float minDistance = std::max<float>((125.f / totalDensity), 2.0f);
+		//distance lower than 2 causes objects to overlap and crash
+		
+		while(true)
+		{
+			auto treasurePileInfos = prepareTreasurePile(t);
+			if(treasurePileInfos.empty())
+				break;
+			
+			int value = std::accumulate(treasurePileInfos.begin(), treasurePileInfos.end(), 0, [](int v, const ObjectInfo * oi){return v + oi->value;});
+			
+			auto rmgObject = constuctTreasurePile(treasurePileInfos);
+			if(rmgObject.instances().empty()) //handle incorrect placement
+			{
+				restoreZoneLimits(treasurePileInfos);
+				continue;
+			}
+			
+			//guard treasure pile
+			bool guarded = isGuardNeededForTreasure(value);
+			if(guarded)
+				guarded = manager.addGuard(rmgObject, value);
+			
+			int3 pos;
+			auto possibleArea = zone.areaPossible();
+			
+			auto path = rmg::Path::invalid();
+			if(guarded)
+			{
+				path = manager.placeAndConnectObject(possibleArea, rmgObject, [this, &rmgObject, &minDistance, &manager](const int3 & tile)
+				{
+					auto ti = map.getTile(tile);
+					if(ti.getNearestObjectDistance() < minDistance)
+						return -1.f;
+					
+					auto guardedArea = rmgObject.instances().back()->getAccessibleArea();
+					auto areaToBlock = rmgObject.getAccessibleArea(true);
+					areaToBlock.subtract(guardedArea);
+					if(areaToBlock.overlap(zone.freePaths()) || areaToBlock.overlap(manager.getVisitableArea()))
+						return -1.f;
+					
+					return 1.f;
```

Comment:
Naaah, the real object distance should be returned so you actually pick the most distant location.

As it is now, you still iterate through all the tiles but use the first matching one anyway, am I correct?

Reply:
right, placeAndConnectObject has optimizer argument which defines behavior: true is to look for best position (maximise functional), false is to break on first positive value of the functional. I use false here in order for performance reasoning

Reply:
First positive value leads to non-uniform distribution. Objects are placed from the top to bottom (and from left to right?), which leads to dense top part of the zone and often empty bottom.

Also, notice that placing object at most distant location leads to convergence since distances around objects get reduced quickly.

I already figured out an algorithm to choose distances quickly with a priority queue, but this can wait for another day.

Reply:
That's true, you pointed on that issue with template marathon. But even with enabled optimiser it will create regular structures which may be not so nice. I think maybe adding some random value would be ok.
Do you think it's regression relatively previous algorithm?

Reply:
Certainly it's a regression if zones end up empty. In fact it's the major bug in current build.

The idea was to place large treasures at large distances from each other first, then fill the gaps with smaller treasures and higher density.

Reply:
I found the problem and made a fix. It was related to fractalization algorithm, not the treasure placer.